lib/metrics/context.js

```diff
@@ -6,43 +6,178 @@
 
 var isA = require('joi')
 var HEX = require('../routes/validators').HEX_STRING
+var P = require('../promise')
+var Memcached = require('memcached')
+P.promisifyAll(Memcached.prototype)
 
-module.exports = {
-  schema: isA.object({
-    flowId: isA.string().length(64).regex(HEX).optional(),
-    flowBeginTime: isA.number().integer().positive().optional(),
-    context: isA.string().optional(),
-    entrypoint: isA.string().optional(),
-    migration: isA.string().optional(),
-    service: isA.string().optional(),
-    utmCampaign: isA.string().optional(),
-    utmContent: isA.string().optional(),
-    utmMedium: isA.string().optional(),
-    utmSource: isA.string().optional(),
-    utmTerm: isA.string().optional()
-  }).and('flowId', 'flowBeginTime').optional(),
-
-  add: function (data, metadata, doNotTrack) {
-    if (metadata) {
-      data.time = Date.now()
-      data.flow_id = metadata.flowId
-      data.flow_time = calculateFlowTime(data.time, metadata.flowBeginTime)
-      data.context = metadata.context
-      data.entrypoint = metadata.entrypoint
-      data.migration = metadata.migration
-      data.service = metadata.service
-
-      if (! doNotTrack) {
-        data.utm_campaign = metadata.utmCampaign
-        data.utm_content = metadata.utmContent
-        data.utm_medium = metadata.utmMedium
-        data.utm_source = metadata.utmSource
-        data.utm_term = metadata.utmTerm
+var SCHEMA = isA.object({
+  flowId: isA.string().length(64).regex(HEX).optional(),
+  flowBeginTime: isA.number().integer().positive().optional(),
+  context: isA.string().optional(),
+  entrypoint: isA.string().optional(),
+  migration: isA.string().optional(),
+  service: isA.string().optional(),
+  utmCampaign: isA.string().optional(),
+  utmContent: isA.string().optional(),
+  utmMedium: isA.string().optional(),
+  utmSource: isA.string().optional(),
+  utmTerm: isA.string().optional()
+}).and('flowId', 'flowBeginTime').optional()
+
+var NOP = function () {
+  return P.resolve()
+}
+
+var NULL_MEMCACHED = {
+  delAsync: NOP,
+  getAsync: NOP,
+  setAsync: NOP
+}
+
+module.exports = function (log, config) {
+  var _memcached
+
+  return {
+    schema: SCHEMA,
+    save: save,
+    copy: copy,
+    remove: remove
```

Comment:
And now for my customary "what should we name this bikeshed?" rambling...feel free to consider or ignore as you please, I'm fully aware I can get a little obsessive about naming things :-)

Something about reading `metricsContext.copy(...)` seemed hard to grok for me, it wasn't obvious exactly what was being copied or to where.  IIUC this method looks in several different places to find the metrics context data, combines any so found into a single data bundle, and returns it for use by the caller.  I wonder if a name like `extract()` or `gather()` might make it more obvious what to expect when you call this method.

I also wonder whether `save` and `remove` need something in the name to indicate that they're for one-time event-related data; perhaps something like `stashEvents` and `dropEvent`?